### PR TITLE
Story3

### PR DESCRIPTION
**Pull request by cbvvrd**
Story 3 completed, pls merge ?

### CODE REVIEW

File: app/Http/Controllers/BookingAPIController.php [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+use App\Models\Room;
+use Carbon\CarbonPeriod;
+use http\Env\Response;
+use Illuminate\Http\Request;
+
+class BookingAPIController extends Controller
+{
+    public function create(Request $request)
+    {
+
```
Comment (ashleycoles) [inline app/Http/Controllers/BookingAPIController.php]: Remove blank line

File: app/Http/Controllers/BookingAPIController.php [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        $request->validate([
+            'room_id' => 'required|integer|exists:rooms,id',
+            'customer' => 'required|string|max:255',
+            'guests' => 'required|integer|min:1',
+            'start' => 'required|date|date_format:Y-m-d|after_or_equal:today',
+            'end' => 'required|date|date_format:Y-m-d|after:start',
```
Comment (ashleycoles) [inline app/Http/Controllers/BookingAPIController.php:21]: I had no idea these after options existed - that's cool! 

It will mean that your API doesn't return what the front-end is expecting when the dates are the wrong way around, so the front-end probably won't display that error message correctly.

Honestly it's up to you whether or not that's an issue - personally I think this is a really cool solution that I'm very happy to accept.

File: app/Http/Controllers/BookingAPIController.php [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        $bookingConflict = Booking::where('room_id', $room->id)
+            ->where(function ($query) use ($startDate, $endDate) {
+                $query->whereBetween('start', [$startDate, $endDate])
+                    ->orWhereBetween('end', [$startDate, $endDate])
+                    ->orWhere(function ($query) use ($startDate, $endDate) {
+                        $query->where('start', '<=', $startDate)
+                            ->where('end', '>=', $endDate);
+                    });
+            })
+            ->exists();
```
Comment (ashleycoles) [inline app/Http/Controllers/BookingAPIController.php]: This would be really nice to have as it's own service. 

Partly because it hides the complicated query behind a method, and also because this may well be something that needs to be repeated in the future (although probably not for these stories)

File: app/Http/Controllers/BookingAPIController.php [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+        if ($room->min_capacity > $request->guests OR  $room->max_capacity < $request->guests) {
+            return response()->json([
+                'message' => "The {$room->name} room can only accommodate between {$room->min_capacity} and {$room->max_capacity} guests",
+            ],400);
+        }
+        elseif ($bookingConflict)
+        {
```
Comment (ashleycoles) [inline app/Http/Controllers/BookingAPIController.php]: Bracket should be on the same line as elseif

File: app/Http/Controllers/BookingAPIController.php [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+                    });
+            })
+            ->exists();
+
+        if ($room->min_capacity > $request->guests OR  $room->max_capacity < $request->guests) {
+            return response()->json([
+                'message' => "The {$room->name} room can only accommodate between {$room->min_capacity} and {$room->max_capacity} guests",
+            ],400);
+        }
+        elseif ($bookingConflict)
```
Comment (ashleycoles) [inline app/Http/Controllers/BookingAPIController.php]: You should make this an if instead of an else if. 

Since the first if returns, nothing else will happen if the guest number is wrong

File: app/Http/Controllers/BookingAPIController.php [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        {
+                return response()->json([
+                    'message' => "The {$room->name} room is unavailable for the chosen dates."
+                ], 400);
+        }
+        else
```
Comment (ashleycoles) [inline app/Http/Controllers/BookingAPIController.php]: You can drop this else now.

if (guests are wrong check) {
    return error response
}

if (booking conflict) {
    return error response
}

make the booking 

return success response

Comment (ashleycoles) [reply app/Http/Controllers/BookingAPIController.php]: This is a practice often called early returning - it can significantly reduce the complexity of the code by having small individual if statements vs a big if/elseif/else (that will inevitably get more complex in the future) that can be quite hard to follow

File: app/Http/Controllers/BookingAPIController.php (hunk 3 above)
Comment (ashleycoles) [reply app/Http/Controllers/BookingAPIController.php]: Would suggest creating a service called `BookingService` (in app/Services - you need to make that folder). Never know what other stuff you might end up putting in it.

File: app/Http/Controllers/BookingAPIController.php [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+                $query->whereBetween('start', [$startDate, $endDate])
+                    ->orWhereBetween('end', [$startDate, $endDate])
+                    ->orWhere(function ($query) use ($startDate, $endDate) {
+                        $query->where('start', '<=', $startDate)
+                            ->where('end', '>=', $endDate);
+                    });
+            })
+            ->exists();
+
+        if ($room->min_capacity > $request->guests OR  $room->max_capacity < $request->guests) {
```
Comment (ashleycoles) [inline app/Http/Controllers/BookingAPIController.php]: You could also put this condition into the `BookingService` - that would improve readability in this file.

If you had something that looked like this instead:

```
if (BookingService::guestsAreValid($room, $request->guests) {
....
}

```

It makes it more obvious what this if statement is doing. And again, this may well be a check that you'd need to repeat in the future.

File: tests/Feature/BookingTest.php [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+use Illuminate\Testing\Fluent\AssertableJson;
+use Tests\TestCase;
+
+class BookingTest extends TestCase
+{
+    use DatabaseMigrations;
+
+    public function test_bookRoom_success(): void
+    {
+        Booking::factory()->create();
```
Comment (ashleycoles) [inline tests/Feature/BookingTest.php]: Any need to make a booking for this test?

File: app/Http/Controllers/BookingAPIController.php [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+            $booking->room_id = $request->room_id;
+            $booking->customer = $request->customer;
+            $booking->start = $request->start;
+            $booking->end = $request->end;
+
+            $booking->save();
+
+            return response()->json([
+                'message' => "Booking successfully created",
+                'data' => $booking
```
Comment (ashleycoles) [inline app/Http/Controllers/BookingAPIController.php]: As per the docs, all your response here needs is the message and the 201 status code - remove data (and make sure to update your test)

File: tests/Feature/BookingTest.php [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+                ])->whereAllType([
+                    'room_id' => 'integer',
+                    'customer' => 'string',
+                    'start' => 'string',
+                    'end' => 'string',
+                    'updated_at' => 'string|null',
+                    'created_at' => 'string|null',
+                    'id' => 'integer'
+                ]);
+            });
```
Comment (ashleycoles) [inline tests/Feature/BookingTest.php]: As per my comment in the controller, this data shouldn't be in this response.

Comment (ashleycoles) [reply tests/Feature/BookingTest.php]: What you should do however, is a database assertion to make sure the new booking was actually saved to the test DB

File: tests/Feature/BookingTest.php [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+    {
+        $testData = [];
+
+        $response = $this->postJson('/api/bookings', $testData);
+
+        $response->assertStatus(422)
+            ->assertInvalid(['room_id', 'customer', 'guests', 'start', 'end'])
+            ->assertJson(function (AssertableJson $json){
+                $json->hasAll('message', 'errors');
+        });
```
Comment (ashleycoles) [inline tests/Feature/BookingTest.php]: No need for this assertJson - assertInvalid is already doing this automatically

File: tests/Feature/BookingTest.php [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+            "room_id" => 1,
+            "customer" => "Test customer",
+            "guests" => 2,
+            "start" => "2026-12-25",
+            "end" => "2026-12-01"
+        ];
+
+        $response = $this->postJson('/api/bookings', $testData);
+
+        $response->assertStatus(422) // Is 422 ok here???? //
```
Comment (ashleycoles) [inline tests/Feature/BookingTest.php]: Yeah 422 is ok given your current solution for checking the dates. The api docs does ask for a 400 with a custom message (one not generated by validation), but I don't mind if you want to stick with what you have now

File: tests/Feature/BookingTest.php [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+            ->assertJson(function (AssertableJson $json){
+                $json->hasAll(['message', 'errors']);
+            });
```
Comment (ashleycoles) [inline tests/Feature/BookingTest.php]: You should use assertInvalid here instead

File: tests/Feature/RoomTest.php [hunk 14]
```diff
@@ -17,6 +18,7 @@ class RoomTest extends TestCase
     public function test_getAllRooms_success(): void
     {
         Room::factory()->create();
+        Type::factory()->create();
```
Comment (ashleycoles) [inline tests/Feature/RoomTest.php]: You probably don't need this anymore since your room factory creates it's own type automatically

File: tests/Feature/BookingTest.php [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+
+        $testData = [
+            "room_id" => 1,
+            "customer" => "Test customer",
+            "guests" => 2,
+            "start" => "2024-12-15",
+            "end" => "2024-12-25"
+        ];
+
+        $response = $this->post('/api/bookings', $testData);
```
Comment (ashleycoles) [inline tests/Feature/BookingTest.php]: You should use postJson() instead

File: tests/Feature/BookingTest.php [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+
+        $testData = [
+            "room_id" => 1,
+            "customer" => "Test customer",
+            "guests" => 500,
+            "start" => "2024-12-15",
+            "end" => "2024-12-25"
+        ];
+
+        $response = $this->post('/api/bookings', $testData);
```
Comment (ashleycoles) [inline tests/Feature/BookingTest.php:68]: postJson() instead

File: app/Http/Controllers/BookingAPIController.php [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+            'guests' => 'required|integer|min:1',
+            'start' => 'required|date|date_format:Y-m-d|after_or_equal:today',
+            'end' => 'required|date|date_format:Y-m-d|after:start',
+        ]);
+
+        $room = Room::find($request->room_id);
+        $startDate = $request->start;
+        $endDate = $request->end;
+
+
```
Comment (ashleycoles) [inline app/Http/Controllers/BookingAPIController.php:25]: Remove extra blank line

File: app/Http/Controllers/BookingAPIController.php [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+            'room_id' => 'required|integer|exists:rooms,id',
+            'customer' => 'required|string|max:255',
+            'guests' => 'required|integer|min:1',
+            'start' => 'required|date|date_format:Y-m-d|after_or_equal:today',
+            'end' => 'required|date|date_format:Y-m-d|after:start',
+        ]);
+
+        $room = Room::find($request->room_id);
+        $startDate = $request->start;
+        $endDate = $request->end;
```
Comment (ashleycoles) [inline app/Http/Controllers/BookingAPIController.php]: I would actually suggest not using these variables, and instead using $request directly when you need the dates. 

While having the variables is less typing, it does make it less obvious what the $startDate and $endDate actually are - whereas having $request->start makes it super clear that it's the start that the user sent in their request

**Review by ashleycoles — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/BookingAPIController.php (hunk 2 above)
Comment (Jodepher) [reply app/Http/Controllers/BookingAPIController.php:21]: Cool solution stays cool 😎

File: app/Http/Controllers/BookingAPIController.php (hunk 3 above)
Comment (Jodepher) [reply app/Http/Controllers/BookingAPIController.php]: Service has been added 🤯

File: tests/Feature/BookingTest.php (hunk 11 above)
Comment (Jodepher) [reply tests/Feature/BookingTest.php]: The assertInvalid is part of a different test - do we need it here as well?

File: tests/Feature/BookingTest.php (hunk 13 above)
Comment (Jodepher) [reply tests/Feature/BookingTest.php]: Please help

**Review by ashleycoles — APPROVED** [review]
